src/relstorage/cache/storage_cache.py

```diff
@@ -772,11 +773,63 @@ def after_poll(self, cursor, prev_tid_int, new_tid_int, changes):
 
             self.__poll_replace_checkpoints(cursor, global_checkpoints, new_tid_int)
 
-        if not global_checkpoints_in_future and len(self.delta_after0) >= self.delta_size_limit:
+        if not global_checkpoints_in_future and self._should_suggest_shifted_checkpoints():
             self._suggest_shifted_checkpoints()
 
         return changes
 
+    #: By default, a 70% chance when we're full.
+    CP_REPLACEMENT_CHANCE_WHEN_FULL = float(
+        os.environ.get(
+            'RELSTORAGE_CP_REPLACEMENT_CHANCE_WHEN_FULL',
+            "0.7"
+        )
+    )
+
+    #: If we're just close, a 20% chance.
+    CP_REPLACEMENT_CHANCE_WHEN_CLOSE = float(
+        os.environ.get(
+            'RELSTORAGE_CP_REPLACEMENT_CHANCE_WHEN_CLOSE',
+            "0.2"
+        )
+    )
+
+    #: Start considering that we're close when we're 80% full.
+    CP_REPLACEMENT_BEGIN_CONSIDERING_PERCENT = float(
+        os.environ.get(
+            'RELSTORAGE_CP_REPLACEMENT_BEGIN_CONSIDERING_PERCENT',
+            "0.8"
+        )
+    )
+
+    def _should_suggest_shifted_checkpoints(self, _random=random.random):
+        """
+        Take the size of the checkpoints and our thresholds into account
+        and determine whether we should try to replace them.
+        """
+
+        delta_size = len(self.delta_after0)
+        limit = self.delta_size_limit
+
+        if delta_size < (limit * self.CP_REPLACEMENT_BEGIN_CONSIDERING_PERCENT):
+            return False
+
+        logger.debug(
+            "Rolling dice to see if checkpoints should shift. len(delta_after0)=%d",
+            delta_size
+        )
+
+        if delta_size >= limit:
+            chances = self.CP_REPLACEMENT_CHANCE_WHEN_FULL
+            when_dice_not_used = True
+        else:
+            chances = self.CP_REPLACEMENT_CHANCE_WHEN_CLOSE
+            when_dice_not_used = False
+        if chances < 1 and _random() >= chances:
+            return False
+        return when_dice_not_used
```

Comment:
Do we ever return True here if we're in the 'close' state?

Reply:
Excellent point,thanks. I shouldn’t have skipped the unit test.